src/automatic.ts

```diff
@@ -28,60 +28,158 @@ import {Server} from './server';
-    // It's important that we execute each benchmark iteration in a new tab.
-    // At least in Chrome, each tab corresponds to process which shares some
-    // amount of cached V8 state which can cause significant measurement
-    // effects. There might even be additional interaction effects that
-    // would require an entirely new browser to remove, but experience in
-    // Chrome so far shows that new tabs are neccessary and sufficient.
-    const driver = await makeDriver(browser);
-    const tabs = await driver.getAllWindowHandles();
-    // We'll always launch new tabs from this initial blank tab.
-    const initialTabHandle = tabs[0];
-    browsers.set(sig, {name: browser.name, driver, initialTabHandle});
   }
 
-  const runSpec = async(spec: BenchmarkSpec): Promise<BenchmarkResult> => {
+  private async takeMinimumSamples() {
+    // Always collect our minimum number of samples.
+    const {config, specs, bar, specResults} = this;
+    const numRuns = specs.length * config.sampleSize;
+    let run = 0;
+    for (let sample = 0; sample < config.sampleSize; sample++) {
+      for (const spec of specs) {
+        bar.tick(0, {
+          status: `${++run}/${numRuns} ${benchmarkOneLiner(spec)}`,
+        });
+        specResults.get(spec)!.push(await this.takeSample(spec));
+        if (bar.curr === bar.total - 1) {
+          // Note if we tick with 0 after we've completed, the status is
+          // rendered on the next line for some reason.
+          bar.tick(1, {status: 'done'});
+        } else {
+          bar.tick(1);
+        }
+      }
+    }
+  }
+
+  private async takeAdditionalSamples() {
+    const {config, specs, specResults} = this;
+    if (config.timeout > 0) {
+      console.log();
+      const timeoutMs = config.timeout * 60 * 1000;  // minutes -> millis
+      const startMs = Date.now();
+      let run = 0;
+      let sample = 0;
+      let elapsed = 0;
+      while (true) {
+        if (horizonsResolved(this.makeResults(), config.horizons)) {
+          console.log();
+          break;
+        }
+        if (elapsed >= timeoutMs) {
+          this.hitTimeout = true;
+          break;
+        }
+        // Run batches of 10 additional samples at a time for more presentable
+        // sample sizes, and to nudge sample sizes up a little.
+        for (let i = 0; i < 10; i++) {
+          sample++;
+          for (const spec of specs) {
+            run++;
+            elapsed = Date.now() - startMs;
+            const remainingSecs =
+                Math.max(0, Math.round((timeoutMs - elapsed) / 1000));
+            const mins = Math.floor(remainingSecs / 60);
+            const secs = remainingSecs % 60;
+            process.stdout.write(
+                `\r${spinner[run % spinner.length]} Auto-sample ${sample} ` +
+                `(timeout in ${mins}m${secs}s)` + ansi.erase.inLine(0));
+            specResults.get(spec)!.push(await this.takeSample(spec));
+          }
+        }
+      }
```

Comment:
Consider flipping this conditional to leave the main path through the code unindented

```suggestion
  private async takeAdditionalSamples() {
    const {config, specs, specResults} = this;
    if (config.timeout <= 0) {
      return;
    }
    console.log();
    const timeoutMs = config.timeout * 60 * 1000;  // minutes -> millis
    const startMs = Date.now();
    let run = 0;
    let sample = 0;
    let elapsed = 0;
    while (true) {
      if (horizonsResolved(this.makeResults(), config.horizons)) {
        console.log();
        break;
      }
      if (elapsed >= timeoutMs) {
        this.hitTimeout = true;
        break;
      }
      // Run batches of 10 additional samples at a time for more presentable
      // sample sizes, and to nudge sample sizes up a little.
      for (let i = 0; i < 10; i++) {
        sample++;
        for (const spec of specs) {
          run++;
          elapsed = Date.now() - startMs;
          const remainingSecs =
            Math.max(0, Math.round((timeoutMs - elapsed) / 1000));
          const mins = Math.floor(remainingSecs / 60);
          const secs = remainingSecs % 60;
          process.stdout.write(
            `\r${spinner[run % spinner.length]} Auto-sample ${sample} ` +
            `(timeout in ${mins}m${secs}s)` + ansi.erase.inLine(0));
          specResults.get(spec)!.push(await this.takeSample(spec));
        }
      }
    }
  }
```

Reply:
Done.